src/Microsoft.Azure.NotificationHubs/Messaging/EntityDescriptionSerializer.cs

```diff
@@ -160,7 +164,10 @@ public string Serialize(EntityDescription description)
                 description = new GcmTemplateRegistrationDescription((FcmTemplateRegistrationDescription) description);
             }
 
-            var serializer = GetSerializer(description.GetType().Name);
+            var serializatorType = description.GetType().Name == typeof(NotificationHubJob).Name ? 
```

Comment:
This change to exclude the `NotificationHubJob` type is too fragile and we should be using the .NET type system instead of comparing type name strings.

In the serialization case, we want to serialize ALL `RegistrationDescription` instances (including derived classes) using the `RegistrationDescription` serializer. This is to maintain compatibility with the existing SDK serialization logic. When serializing, we always know the .NET type being serialized. The type "name" is less interesting.

In the deserialization case, we want to support either the RegistrationDescription serialized format, or the derived class serialized format. When deserializing, we are trying to detect the proper serializer based on the XML element name, so getting the serializer by type "name" is appropriate.

There is also some duplication between the two `Serialize` methods. I think we can improve this code.

For the two `Serialize` methods, one serializes to (and returns) a `string`, while the other serializes to an `XmlWriter`. These can just be the same implementation, with the `string` version creating an `XmlWriter` and calling the `XmlWriter` version.

Once we've removed the duplication between the two `Serialize` methods, we can then simplify the logic for choosing the correct serializer type. That logic should be based on .NET type. Something like:
```
DataContractSerializer serializer;
if (description is RegistrationDescription)
{
    serializer = GetSerializer(typeof(RegistrationDescription).Name);
}
else
{
    serializer = GetSerializer(description.GetType().Name);
}
```
I still don't like that we use the type name to determine the correct serializer in the `Serialize` code path but, in the interest of code reuse, it's not worth changing right now.

Reply:
@brannon 
I fixed that